### PR TITLE
Stable hats

### LINKED ISSUE #432: Make hat allocations more stable
### The problem

Today, we don't make any effort to use the same hats for the same tokens when we refresh the hat map.  This behaviour can lead to problems where

- the hat map can shift while you're thinking / formulating a command,
- it can be distracting
- you cannot start issuing a new cursorless command while the previous one is running
- race conditions with scrolling / edits while you're starting to say a command

The issue was improved by #318, but that only works for single phrases, not continuity between phrases

### The solution

We could treat hats more like a resource, the way a memory allocator treats memory blocks.  Then when we go to reallocate the hat map, we only allocate new hats, or hats that have been freed from a deletion / scroll

Note: a lot of implementation might overlap with #308, and arguably this issue is a better solution to the problem that #308 attempts to solve, though #308 would allow frozen sections to persist even when off screen

Fwiw, here is some initial work: https://github.com/cursorless-dev/cursorless/pull/1225 / https://github.com/cursorless-dev/cursorless/pull/1225/commits/a4802ea3851d8552075db0e56353c9ecab666eb1

### PR DESCRIPTION
**Pull request by pokey**
- Fixes #432 

## Checklist

- [ ] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [ ] Add test for case where enabled hat styles change, both adding or removing an enabled hat style
- [ ] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [ ] I have not broken the cheatsheet

### CODE REVIEW

**Comment by AndreasArvidsson** [comment]
After some testing I really like this change. Nice work!

File: src/core/IndividualHatMap.ts [hunk 1]
```diff
@@ -2,21 +2,32 @@ import { getKey, TextDocument } from "@cursorless/common";
...
+import { FullRangeInfo } from "../typings/updateSelections";
+import { getMatcher } from "../libs/cursorless-engine/tokenizer";
 
 export interface ReadOnlyHatMap {
-  getEntries(): [string, Token][];
+  getEntries(): readonly [string, Token][];
   getToken(hatStyle: HatStyleName, character: string): Token;
 }
 
+interface LiveToken extends Token, FullRangeInfo {}
```
Comment (pokey) [inline src/core/IndividualHatMap.ts:14]: docstring

File: src/libs/common/ide/types/HatStability.ts [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+  lowRounded = 1,
+
+  /**
+   * Only try to reuse an old hat when doing so wouldn't cause a token's hat
+   * penalty to go from <2 to >=2
+   */
+  lowThresholded = 2,
+
+  /**
+   * TODO: Document
```
Comment (pokey) [inline src/libs/common/ide/types/HatStability.ts:21]: do this

File: src/util/allocateHats/allocateHats.ts [hunk 3]
```diff
@@ -0,0 +1,132 @@
...
+
+  /**
+   * A map from graphemes to the remaining hat styles that have not yet been
+   * used for that grapheme
+   */
+  const graphemeRemainingHatCandidates = new DefaultMap<string, HatStyleMap>(
+    () => clone(enabledHatStyles),
+  );
+
+  return tokens
```
Comment (pokey) [inline src/util/allocateHats/allocateHats.ts:53]: comment above this

File: src/util/allocateHats/allocateHats.ts [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+  hatRange: Range;
+}
+
+export interface HatCandidate {
+  grapheme: Grapheme;
+  style: HatStyleName;
+  penalty: number;
+}
+
+export function allocateHats(
```
Comment (pokey) [inline src/util/allocateHats/allocateHats.ts:29]: docstring

File: src/util/allocateHats/chooseTokenHat.ts [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+import { HatMetrics, HatMetric } from "./HatMetrics";
+
+/**
+ * IMPORTANT: This function assumes that all tokens with a lower rank than the given
+ * token have already been assigned hats.
+ *
+ * Picks the character with minimum color such that the next token that contains
+ * that character is as far away as possible.
+ *
+ * TODO: Could be improved by ignoring subsequent tokens that also contain
```
Comment (pokey) [inline src/util/allocateHats/chooseTokenHat.ts:14]: should be FIXME not TODO

File: src/util/allocateHats/chooseTokenHat.ts [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+  }
+}
+
+function getHatPenaltyEquivalenceClassFn(
+  hatStability: HatStability,
+): HatMetric {
+  switch (hatStability) {
+    case HatStability.low:
+      return ({ penalty }) => penalty;
+    case HatStability.lowRounded:
```
Comment (pokey) [inline src/util/allocateHats/chooseTokenHat.ts:109]: lowFloored

File: src/util/allocateHats/chooseTokenHat.ts [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+      return ({ penalty }) => Math.floor(penalty);
+    case HatStability.lowThresholded:
+    case HatStability.thresholded:
+      return ({ penalty }) => (penalty < 2 ? 0 : 1);
```
Comment (pokey) [inline src/util/allocateHats/chooseTokenHat.ts:113]:
```suggestion
      return ({ penalty }) => (penalty < 2 ? 0 : penalty);
```

File: src/util/allocateHats/chooseTokenHat.ts [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+      : candidates.find(
+          (hat) =>
+            hat.grapheme.text === oldTokenHat.grapheme &&
+            hat.style === oldTokenHat.hatStyle,
+        );
+
+  if (oldCandidate != null) {
+    if (hatStability >= HatStability.high) {
+      return oldCandidate;
+    }
```
Comment (pokey) [inline src/util/allocateHats/chooseTokenHat.ts:50]: remove special case

File: src/util/allocateHats/maxByMultiple.ts [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+ * @example maxByMultiple([{a: 1, b: 1}, {a: 1, b: 2}], [({a}) => a, ({b}) =>
+ * b]) === {a: 1, b: 2}
+ *
+ * @param arr The array to find the max value of
+ * @param fns A list of functions that return a number for each item in the
+ * array
+ * @returns The item in the array that has the maximum value, or undefined if
+ * array is empty or all items are removed
+ */
+export function maxByMultiple<T>(
```
Comment (pokey) [inline src/util/allocateHats/maxByMultiple.ts:26]: maybe `maxByFirstDiffering`?

File: src/util/allocateHats/HatMetrics.ts [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+  getHatOldTokenRank: HatMetric = ({ grapheme: { text: grapheme }, style }) => {
+    const hatOldTokenRank = this.hatOldTokenRanks.get({
+      grapheme,
+      hatStyle: style,
+    });
+
+    return hatOldTokenRank == null
+      ? Infinity
+      : this.hatStability === HatStability.strict
+      ? NaN
```
Comment (pokey) [inline src/util/allocateHats/HatMetrics.ts:41]: maybe just do this via a filtering step

File: src/util/allocateHats/chooseTokenHat.ts [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+function getHatTheftMetrics(
+  hatStability: HatStability,
+  metrics: HatMetrics,
+): HatMetric[] {
+  switch (hatStability) {
+    case HatStability.low:
+    case HatStability.lowRounded:
+    case HatStability.lowThresholded:
+    case HatStability.high:
+      return [metrics.getNegativePenalty, metrics.getHatOldTokenRank];
```
Comment (pokey) [inline src/util/allocateHats/chooseTokenHat.ts:87]: comment

File: package.json [hunk 12]
```diff
@@ -687,6 +687,31 @@
...
+          "default": "low",
+          "enum": [
+            "low",
+            "lowRounded",
+            "lowThresholded",
+            "thresholded",
+            "high",
+            "highThresholded",
+            "higher",
+            "strict"
```
Comment (pokey) [inline package.json:703]: try to orthogonalise into a threshold and then low, medium, high or something like that

File: src/libs/common/util/CompositeKeyMap.ts [hunk 13]
```diff
@@ -0,0 +1,32 @@
+// Based on https://stackoverflow.com/a/54523103
+export default class CompositeKeyMap<K, V> {
```
Comment (pokey) [inline src/libs/common/util/CompositeKeyMap.ts:2]: docstring

File: src/libs/common/util/DefaultMap.ts [hunk 14]
```diff
@@ -0,0 +1,19 @@
+// Based on https://aaronmoat.com/implementing-pythons-defaultdict-in-javascript/
+export default class DefaultMap<K, V> extends Map<K, V> {
```
Comment (pokey) [inline src/libs/common/util/DefaultMap.ts:2]: docstring

File: src/testUtil/plainObjectToTarget.ts [hunk 15]
```diff
@@ -21,6 +17,7 @@ import type { Target } from "../typings/target.types";
  * @param plainObject A plain object describing a `Target`
  * @returns A `Target` constructed from the given plain object
  */
+
```
Comment (pokey) [inline src/testUtil/plainObjectToTarget.ts:20]:
```suggestion
```

**Review by AndreasArvidsson — APPROVED** [review]

**Pull request merged**